agent/xds/listeners.go

```diff
@@ -0,0 +1,286 @@
+			if err != nil {
+				return l, err
+			}
+		}
+	}
+
+	if l == nil {
+		// No user config, use default listener
+		addr := cfgSnap.Address
+		if addr == "" {
+			addr = "0.0.0.0"
+		}
+		l = makeListener(PublicListenerName, addr, cfgSnap.Port)
+		tcpProxy, err := makeTCPProxyFilter("public_listener", LocalAppClusterName)
+		if err != nil {
+			return l, err
+		}
+		// Setup TCP proxy for now. We inject TLS and authz below for both default
+		// and custom config cases.
+		l.FilterChains = []envoylistener.FilterChain{
+			{
+				Filters: []envoylistener.Filter{
+					tcpProxy,
+				},
+			},
+		}
+	}
+
+	err = injectConnectFilters(cfgSnap, token, l)
+	return l, err
+}
+
+func makeUpstreamListener(u *structs.Upstream) (proto.Message, error) {
+	if listenerJSONRaw, ok := u.Config["envoy_listener_json"]; ok {
+		if listenerJSON, ok := listenerJSONRaw.(string); ok {
+			return makeListenerFromUserConfig(listenerJSON)
+		}
+	}
+	addr := u.LocalBindAddress
+	if addr == "" {
+		addr = "127.0.0.1"
+	}
+	l := makeListener(u.Identifier(), addr, u.LocalBindPort)
+	tcpProxy, err := makeTCPProxyFilter(u.Identifier(), u.Identifier())
+	if err != nil {
+		return l, err
+	}
+	l.FilterChains = []envoylistener.FilterChain{
+		{
+			Filters: []envoylistener.Filter{
+				tcpProxy,
+			},
+		},
+	}
+	return l, nil
+}
+
+func makeTCPProxyFilter(name, cluster string) (envoylistener.Filter, error) {
+	cfg := &envoytcp.TcpProxy{
+		StatPrefix: name,
+		Cluster:    cluster,
+	}
+	return makeFilter("envoy.tcp_proxy", cfg)
+}
+
+func makeExtAuthFilter(token string) (envoylistener.Filter, error) {
+	cfg := &extauthz.ExtAuthz{
+		StatPrefix: "connect_authz",
+		GrpcService: &envoycore.GrpcService{
+			// Attach token header so we can authorize the callbacks. Technically
+			// authorize is not really protected data but we locked down the HTTP
+			// implementation to need service:write and since we have the token that
+			// has that it's pretty reasonable to set it up here.
+			InitialMetadata: []*envoycore.HeaderValue{
+				&envoycore.HeaderValue{
+					Key:   "x-consul-token",
+					Value: token,
+				},
+			},
+			TargetSpecifier: &envoycore.GrpcService_EnvoyGrpc_{
+				EnvoyGrpc: &envoycore.GrpcService_EnvoyGrpc{
+					ClusterName: LocalAgentClusterName,
+				},
+			},
+		},
+		FailureModeAllow: false,
+	}
+	return makeFilter("envoy.ext_authz", cfg)
+}
+
+func makeFilter(name string, cfg proto.Message) (envoylistener.Filter, error) {
+	// Ridiculous dance to make that pbstruct into types.Struct by... encoding it
+	// as JSON and decoding again!!
+	cfgStruct, err := util.MessageToStruct(cfg)
+	if err != nil {
+		return envoylistener.Filter{}, err
+	}
+
+	return envoylistener.Filter{
+		Name:   name,
+		Config: cfgStruct,
+	}, nil
+}
+
+func makeCommonTLSContext(cfgSnap *proxycfg.ConfigSnapshot) *envoyauth.CommonTlsContext {
+	// Concatenate all the root PEMs into one.
+	// TODO(banks): verify this actually works with Envoy (docs are not clear).
```

Comment:
TODO now or later?

Reply:
Ha, well not done yet technically would be a nice thing to get going in a test setup as part of the release cycle.